File: tests/abi/audio_processor_layout.cpp
#include "pluginterfaces/vst/ivstaudioprocessor.h"
#include <cstring>
#include "public.sdk/source/vst/vstaudioprocessoralgo.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Vst::Type), alignof (Steinberg::Vst::Type))
#define PRINT_OFFSET(Type, Field) std::printf (#Type "." #Field " offset %zu\n", offsetof (Steinberg::Vst::Type, Field))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	std::printf ("ComponentFlags.kDistributable %u\n", Steinberg::Vst::kDistributable);
	std::printf ("ComponentFlags.kSimpleModeSupported %u\n", Steinberg::Vst::kSimpleModeSupported);
	std::printf ("SymbolicSampleSizes.kSample32 %d\n", Steinberg::Vst::kSample32);
	std::printf ("SymbolicSampleSizes.kSample64 %d\n", Steinberg::Vst::kSample64);
	std::printf ("ProcessModes.kRealtime %d\n", Steinberg::Vst::kRealtime);
	std::printf ("ProcessModes.kPrefetch %d\n", Steinberg::Vst::kPrefetch);
	std::printf ("ProcessModes.kOffline %d\n", Steinberg::Vst::kOffline);
	std::printf ("kNoTail %u\n", Steinberg::Vst::kNoTail);
	std::printf ("kInfiniteTail %u\n", Steinberg::Vst::kInfiniteTail);
	std::printf ("IProcessContextRequirements.kNeedSystemTime %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedSystemTime);
	std::printf ("IProcessContextRequirements.kNeedContinousTimeSamples %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedContinousTimeSamples);
	std::printf ("IProcessContextRequirements.kNeedProjectTimeMusic %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedProjectTimeMusic);
	std::printf ("IProcessContextRequirements.kNeedBarPositionMusic %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedBarPositionMusic);
	std::printf ("IProcessContextRequirements.kNeedCycleMusic %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedCycleMusic);
	std::printf ("IProcessContextRequirements.kNeedSamplesToNextClock %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedSamplesToNextClock);
	std::printf ("IProcessContextRequirements.kNeedTempo %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedTempo);
	std::printf ("IProcessContextRequirements.kNeedTimeSignature %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedTimeSignature);
	std::printf ("IProcessContextRequirements.kNeedChord %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedChord);
	std::printf ("IProcessContextRequirements.kNeedFrameRate %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedFrameRate);
	std::printf ("IProcessContextRequirements.kNeedTransportState %u\n", Steinberg::Vst::IProcessContextRequirements::kNeedTransportState);

	PRINT_TYPE (ProcessSetup);
	PRINT_OFFSET (ProcessSetup, processMode);
	PRINT_OFFSET (ProcessSetup, symbolicSampleSize);
	PRINT_OFFSET (ProcessSetup, maxSamplesPerBlock);
	PRINT_OFFSET (ProcessSetup, sampleRate);

	PRINT_TYPE (AudioBusBuffers);
	PRINT_OFFSET (AudioBusBuffers, numChannels);
	PRINT_OFFSET (AudioBusBuffers, silenceFlags);
	PRINT_OFFSET (AudioBusBuffers, channelBuffers64);

	PRINT_TYPE (ProcessData);
	PRINT_OFFSET (ProcessData, processMode);
	PRINT_OFFSET (ProcessData, symbolicSampleSize);
	PRINT_OFFSET (ProcessData, numSamples);
	PRINT_OFFSET (ProcessData, numInputs);
	PRINT_OFFSET (ProcessData, numOutputs);
	PRINT_OFFSET (ProcessData, inputs);
	PRINT_OFFSET (ProcessData, outputs);
	PRINT_OFFSET (ProcessData, inputParameterChanges);
	PRINT_OFFSET (ProcessData, outputParameterChanges);
	PRINT_OFFSET (ProcessData, inputEvents);
	PRINT_OFFSET (ProcessData, outputEvents);
	PRINT_OFFSET (ProcessData, processContext);

	Steinberg::Vst::ProcessSetup setup32 {};
	setup32.symbolicSampleSize = Steinberg::Vst::kSample32;
	Steinberg::Vst::ProcessSetup setup64 {};
	setup64.symbolicSampleSize = Steinberg::Vst::kSample64;
	float channel32[4] {};
	double channel64[4] {};
	float* channels32[1] {channel32};
	double* channels64[1] {channel64};
	Steinberg::Vst::AudioBusBuffers buffers32 {};
	buffers32.channelBuffers32 = channels32;
	Steinberg::Vst::AudioBusBuffers buffers64 {};
	buffers64.channelBuffers64 = channels64;
	std::printf ("AudioProcessorAlgo.getChannelBuffersPointer.32 %d\n", Steinberg::Vst::getChannelBuffersPointer (setup32, buffers32) == reinterpret_cast<void**> (channels32));
	std::printf ("AudioProcessorAlgo.getChannelBuffersPointer.64 %d\n", Steinberg::Vst::getChannelBuffersPointer (setup64, buffers64) == reinterpret_cast<void**> (channels64));
	std::printf ("AudioProcessorAlgo.getSampleFramesSizeInBytes.32 %u\n", Steinberg::Vst::getSampleFramesSizeInBytes (setup32, 8));
	std::printf ("AudioProcessorAlgo.getSampleFramesSizeInBytes.64 %u\n", Steinberg::Vst::getSampleFramesSizeInBytes (setup64, 8));
	std::printf ("AudioProcessorAlgo.getChannelMask.0 %llu\n", static_cast<unsigned long long> (Steinberg::Vst::getChannelMask (0)));
	std::printf ("AudioProcessorAlgo.getChannelMask.6 %llu\n", static_cast<unsigned long long> (Steinberg::Vst::getChannelMask (6)));
	std::printf ("AudioProcessorAlgo.getChannelMask.64 %llu\n", static_cast<unsigned long long> (Steinberg::Vst::getChannelMask (64)));
	float src32Ch0[4] {1.f, 2.f, 3.f, 4.f};
	float src32Ch1[4] {5.f, 6.f, 7.f, 8.f};
	float dest32Ch0[6] {};
	float dest32Ch1[6] {};
	float* src32Channels[2] {src32Ch0, src32Ch1};
	float* dest32Channels[2] {dest32Ch0, dest32Ch1};
	Steinberg::Vst::AudioBusBuffers src32 {};
	src32.numChannels = 2;
	src32.channelBuffers32 = src32Channels;
	Steinberg::Vst::AudioBusBuffers dest32 {};
	dest32.numChannels = 2;
	dest32.channelBuffers32 = dest32Channels;
	Steinberg::Vst::Algo::copy32 (&src32, &dest32, 3, 2);
	std::printf ("AudioProcessorAlgo.copy32.dest0.2 %.1f\n", dest32Ch0[2]);
	std::printf ("AudioProcessorAlgo.copy32.dest1.4 %.1f\n", dest32Ch1[4]);
	Steinberg::Vst::Algo::mix32 (src32, dest32, 3);
	std::printf ("AudioProcessorAlgo.mix32.dest0.0 %.1f\n", dest32Ch0[0]);
	std::printf ("AudioProcessorAlgo.mix32.dest1.2 %.1f\n", dest32Ch1[2]);
	Steinberg::Vst::Algo::multiply32 (src32, dest32, 3, 2.f);
	std::printf ("AudioProcessorAlgo.multiply32.dest0.1 %.1f\n", dest32Ch0[1]);
	std::printf ("AudioProcessorAlgo.isSilent32.before %d\n", Steinberg::Vst::Algo::isSilent32 (dest32, 3));
	Steinberg::Vst::Algo::clear32 (&dest32, 3);
	std::printf ("AudioProcessorAlgo.clear32.dest0.1 %.1f\n", dest32Ch0[1]);
	std::printf ("AudioProcessorAlgo.isSilent32.after %d\n", Steinberg::Vst::Algo::isSilent32 (dest32, 3));
	double src64Ch0[3] {1.5, 2.5, 3.5};
	double dest64Ch0[5] {};
	double* src64Channels[1] {src64Ch0};
	double* dest64Channels[1] {dest64Ch0};
	Steinberg::Vst::AudioBusBuffers src64 {};
	src64.numChannels = 1;
	src64.channelBuffers64 = src64Channels;
	Steinberg::Vst::AudioBusBuffers dest64 {};
	dest64.numChannels = 1;
	dest64.channelBuffers64 = dest64Channels;
	Steinberg::Vst::Algo::copy64 (&src64, &dest64, 2, 1);
	std::printf ("AudioProcessorAlgo.copy64.dest0.2 %.1f\n", dest64Ch0[2]);
	Steinberg::Vst::Algo::multiply64 (src64, dest64, 2, 3.0);
	std::printf ("AudioProcessorAlgo.multiply64.dest0.1 %.1f\n", dest64Ch0[1]);
	std::printf ("AudioProcessorAlgo.isSilent64.before %d\n", Steinberg::Vst::Algo::isSilent64 (dest64, 2));
	Steinberg::Vst::Algo::clear64 (&dest64, 2);
	std::printf ("AudioProcessorAlgo.isSilent64.after %d\n", Steinberg::Vst::Algo::isSilent64 (dest64, 2));

	print_iid ("IAudioProcessor", Steinberg::Vst::IAudioProcessor_iid);
	print_iid ("IAudioPresentationLatency", Steinberg::Vst::IAudioPresentationLatency_iid);
	print_iid ("IProcessContextRequirements", Steinberg::Vst::IProcessContextRequirements_iid);
	return 0;
}

File: tools/audio_processor_layout.zig
const std = @import("std");
const base = @import("zig-vst3").pluginterfaces.base.types;
const audio_processor = @import("zig-vst3").pluginterfaces.vst.ivstaudioprocessor;
const audio_processor_algo = @import("zig-vst3").pluginterfaces.vst.vstaudioprocessoralgo;
const events = @import("zig-vst3").pluginterfaces.vst.ivstevents;
const parameter_changes = @import("zig-vst3").pluginterfaces.vst.ivstparameterchanges;
const vsttypes = @import("zig-vst3").pluginterfaces.vst.vsttypes;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try stdout.print("kVstAudioEffectClass {s}\n", .{audio_processor.kVstAudioEffectClass});
    try stdout.print("ComponentFlags.kDistributable {}\n", .{audio_processor.ComponentFlags.kDistributable});
    try stdout.print("ComponentFlags.kSimpleModeSupported {}\n", .{audio_processor.ComponentFlags.kSimpleModeSupported});
    try stdout.print("SymbolicSampleSizes.kSample32 {}\n", .{@intFromEnum(audio_processor.SymbolicSampleSizes.kSample32)});
    try stdout.print("SymbolicSampleSizes.kSample64 {}\n", .{@intFromEnum(audio_processor.SymbolicSampleSizes.kSample64)});
    try stdout.print("ProcessModes.kRealtime {}\n", .{@intFromEnum(audio_processor.ProcessModes.kRealtime)});
    try stdout.print("ProcessModes.kPrefetch {}\n", .{@intFromEnum(audio_processor.ProcessModes.kPrefetch)});
    try stdout.print("ProcessModes.kOffline {}\n", .{@intFromEnum(audio_processor.ProcessModes.kOffline)});
    try stdout.print("kNoTail {}\n", .{audio_processor.kNoTail});
    try stdout.print("kInfiniteTail {}\n", .{audio_processor.kInfiniteTail});
    try stdout.print("IProcessContextRequirements.kNeedSystemTime {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedSystemTime});
    try stdout.print("IProcessContextRequirements.kNeedContinousTimeSamples {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedContinousTimeSamples});
    try stdout.print("IProcessContextRequirements.kNeedProjectTimeMusic {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedProjectTimeMusic});
    try stdout.print("IProcessContextRequirements.kNeedBarPositionMusic {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedBarPositionMusic});
    try stdout.print("IProcessContextRequirements.kNeedCycleMusic {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedCycleMusic});
    try stdout.print("IProcessContextRequirements.kNeedSamplesToNextClock {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedSamplesToNextClock});
    try stdout.print("IProcessContextRequirements.kNeedTempo {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedTempo});
    try stdout.print("IProcessContextRequirements.kNeedTimeSignature {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedTimeSignature});
    try stdout.print("IProcessContextRequirements.kNeedChord {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedChord});
    try stdout.print("IProcessContextRequirements.kNeedFrameRate {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedFrameRate});
    try stdout.print("IProcessContextRequirements.kNeedTransportState {}\n", .{audio_processor.ProcessContextRequirementFlags.kNeedTransportState});

    try printType(stdout, "ProcessSetup", audio_processor.ProcessSetup);
    try printOffset(stdout, "ProcessSetup", "processMode", audio_processor.ProcessSetup, "processMode");
    try printOffset(stdout, "ProcessSetup", "symbolicSampleSize", audio_processor.ProcessSetup, "symbolicSampleSize");
    try printOffset(stdout, "ProcessSetup", "maxSamplesPerBlock", audio_processor.ProcessSetup, "maxSamplesPerBlock");
    try printOffset(stdout, "ProcessSetup", "sampleRate", audio_processor.ProcessSetup, "sampleRate");

    try printType(stdout, "AudioBusBuffers", audio_processor.AudioBusBuffers);
    try printOffset(stdout, "AudioBusBuffers", "numChannels", audio_processor.AudioBusBuffers, "numChannels");
    try printOffset(stdout, "AudioBusBuffers", "silenceFlags", audio_processor.AudioBusBuffers, "silenceFlags");
    try printOffset(stdout, "AudioBusBuffers", "channelBuffers32", audio_processor.AudioBusBuffers, "channelBuffers");
    try printOffset(stdout, "AudioBusBuffers", "channelBuffers64", audio_processor.AudioBusBuffers, "channelBuffers");

    try printType(stdout, "ProcessData", audio_processor.ProcessData);
    try printOffset(stdout, "ProcessData", "processMode", audio_processor.ProcessData, "processMode");
    try printOffset(stdout, "ProcessData", "symbolicSampleSize", audio_processor.ProcessData, "symbolicSampleSize");
    try printOffset(stdout, "ProcessData", "numSamples", audio_processor.ProcessData, "numSamples");
    try printOffset(stdout, "ProcessData", "numInputs", audio_processor.ProcessData, "numInputs");
    try printOffset(stdout, "ProcessData", "numOutputs", audio_processor.ProcessData, "numOutputs");
    try printOffset(stdout, "ProcessData", "inputs", audio_processor.ProcessData, "inputs");
    try printOffset(stdout, "ProcessData", "outputs", audio_processor.ProcessData, "outputs");
    try printOffset(stdout, "ProcessData", "inputParameterChanges", audio_processor.ProcessData, "inputParameterChanges");
    try printOffset(stdout, "ProcessData", "outputParameterChanges", audio_processor.ProcessData, "outputParameterChanges");
    try printOffset(stdout, "ProcessData", "inputEvents", audio_processor.ProcessData, "inputEvents");
    try printOffset(stdout, "ProcessData", "outputEvents", audio_processor.ProcessData, "outputEvents");
    try printOffset(stdout, "ProcessData", "processContext", audio_processor.ProcessData, "processContext");

    try printType(stdout, "IAudioProcessor", audio_processor.IAudioProcessor);
    try printType(stdout, "IAudioPresentationLatency", audio_processor.IAudioPresentationLatency);
    try printType(stdout, "IProcessContextRequirements", audio_processor.IProcessContextRequirements);

    var setup32 = audio_processor.ProcessSetup{ .symbolicSampleSize = @intFromEnum(audio_processor.SymbolicSampleSizes.kSample32) };
    var setup64 = audio_processor.ProcessSetup{ .symbolicSampleSize = @intFromEnum(audio_processor.SymbolicSampleSizes.kSample64) };
    var channel32 = [_]f32{0} ** 4;
    var channel64 = [_]f64{0} ** 4;
    var channels32 = [_][*]f32{&channel32};
    var channels64 = [_][*]f64{&channel64};
    var buffers32 = audio_processor.AudioBusBuffers{ .channelBuffers = .{ .channelBuffers32 = &channels32 } };
    var buffers64 = audio_processor.AudioBusBuffers{ .channelBuffers = .{ .channelBuffers64 = &channels64 } };
    try stdout.print("AudioProcessorAlgo.getChannelBuffersPointer.32 {}\n", .{@intFromBool(audio_processor_algo.getChannelBuffersPointer(&setup32, &buffers32) == @as(?*anyopaque, @ptrCast(&channels32)))});
    try stdout.print("AudioProcessorAlgo.getChannelBuffersPointer.64 {}\n", .{@intFromBool(audio_processor_algo.getChannelBuffersPointer(&setup64, &buffers64) == @as(?*anyopaque, @ptrCast(&channels64)))});
    try stdout.print("AudioProcessorAlgo.getSampleFramesSizeInBytes.32 {}\n", .{audio_processor_algo.getSampleFramesSizeInBytes(&setup32, 8)});
    try stdout.print("AudioProcessorAlgo.getSampleFramesSizeInBytes.64 {}\n", .{audio_processor_algo.getSampleFramesSizeInBytes(&setup64, 8)});
    try stdout.print("AudioProcessorAlgo.getChannelMask.0 {}\n", .{audio_processor_algo.getChannelMask(0)});
    try stdout.print("AudioProcessorAlgo.getChannelMask.6 {}\n", .{audio_processor_algo.getChannelMask(6)});
    try stdout.print("AudioProcessorAlgo.getChannelMask.64 {}\n", .{audio_processor_algo.getChannelMask(64)});
    var src32_ch0 = [_]f32{ 1, 2, 3, 4 };
    var src32_ch1 = [_]f32{ 5, 6, 7, 8 };
    var dest32_ch0 = [_]f32{0} ** 6;
    var dest32_ch1 = [_]f32{0} ** 6;
    var src32_channels = [_][*]f32{ &src32_ch0, &src32_ch1 };
    var dest32_channels = [_][*]f32{ &dest32_ch0, &dest32_ch1 };
    var src32 = audio_processor.AudioBusBuffers{ .numChannels = 2, .channelBuffers = .{ .channelBuffers32 = &src32_channels } };
    var dest32 = audio_processor.AudioBusBuffers{ .numChannels = 2, .channelBuffers = .{ .channelBuffers32 = &dest32_channels } };
    audio_processor_algo.copy32(&src32, &dest32, 3, 2);
    try stdout.print("AudioProcessorAlgo.copy32.dest0.2 {d:.1}\n", .{dest32_ch0[2]});
    try stdout.print("AudioProcessorAlgo.copy32.dest1.4 {d:.1}\n", .{dest32_ch1[4]});
    audio_processor_algo.mix32(&src32, &dest32, 3);
    try stdout.print("AudioProcessorAlgo.mix32.dest0.0 {d:.1}\n", .{dest32_ch0[0]});
    try stdout.print("AudioProcessorAlgo.mix32.dest1.2 {d:.1}\n", .{dest32_ch1[2]});
    audio_processor_algo.multiply32(&src32, &dest32, 3, 2);
    try stdout.print("AudioProcessorAlgo.multiply32.dest0.1 {d:.1}\n", .{dest32_ch0[1]});
    try stdout.print("AudioProcessorAlgo.isSilent32.before {}\n", .{@intFromBool(audio_processor_algo.isSilent32(&dest32, 3, 0))});
    audio_processor_algo.clear32(@ptrCast(&dest32), 3, 1);
    try stdout.print("AudioProcessorAlgo.clear32.dest0.1 {d:.1}\n", .{dest32_ch0[1]});
    try stdout.print("AudioProcessorAlgo.isSilent32.after {}\n", .{@intFromBool(audio_processor_algo.isSilent32(&dest32, 3, 0))});
    var src64_ch0 = [_]f64{ 1.5, 2.5, 3.5 };
    var dest64_ch0 = [_]f64{0} ** 5;
    var src64_channels = [_][*]f64{&src64_ch0};
    var dest64_channels = [_][*]f64{&dest64_ch0};
    var src64 = audio_processor.AudioBusBuffers{ .numChannels = 1, .channelBuffers = .{ .channelBuffers64 = &src64_channels } };
    var dest64 = audio_processor.AudioBusBuffers{ .numChannels = 1, .channelBuffers = .{ .channelBuffers64 = &dest64_channels } };
    audio_processor_algo.copy64(&src64, &dest64, 2, 1);
    try stdout.print("AudioProcessorAlgo.copy64.dest0.2 {d:.1}\n", .{dest64_ch0[2]});
    audio_processor_algo.multiply64(&src64, &dest64, 2, 3);
    try stdout.print("AudioProcessorAlgo.multiply64.dest0.1 {d:.1}\n", .{dest64_ch0[1]});
    try stdout.print("AudioProcessorAlgo.isSilent64.before {}\n", .{@intFromBool(audio_processor_algo.isSilent64(&dest64, 2, 0))});
    audio_processor_algo.clear64(@ptrCast(&dest64), 2, 1);
    try stdout.print("AudioProcessorAlgo.isSilent64.after {}\n", .{@intFromBool(audio_processor_algo.isSilent64(&dest64, 2, 0))});
    var event_list = MockEventList.init();
    var event_context = EventContext{};
    audio_processor_algo.forEachEvent(&event_list.iface, &event_context, collectEvent);
    try stdout.print("AudioProcessorAlgo.foreachEvent.offsetSum {}\n", .{event_context.offset_sum});
    var param_queue = MockParamValueQueue.init(1234);
    var param_context = ParamContext{};
    audio_processor_algo.forEachParamValueQueue(&param_queue.iface, &param_context, collectParamPoint);
    try stdout.print("AudioProcessorAlgo.foreachParam.paramId {}\n", .{param_context.param_id});
    try stdout.print("AudioProcessorAlgo.foreachParam.offsetSum {}\n", .{param_context.offset_sum});
    try stdout.print("AudioProcessorAlgo.foreachParam.valueSum {d:.2}\n", .{param_context.value_sum});
    var last_context = LastParamContext{};
    audio_processor_algo.forEachLastParamValueQueue(&param_queue.iface, &last_context, collectLastParamPoint);
    try stdout.print("AudioProcessorAlgo.foreachLast.offset {}\n", .{last_context.offset});
    try stdout.print("AudioProcessorAlgo.foreachLast.value {d:.2}\n", .{last_context.value});
    var changes = MockParameterChanges.init();
    var changes_context = ChangesContext{};
    audio_processor_algo.forEachParameterChanges(&changes.iface, &changes_context, collectParamQueue);
    try stdout.print("AudioProcessorAlgo.foreachChanges.paramIdSum {}\n", .{changes_context.param_id_sum});

    try printTuid(stdout, "IAudioProcessor", audio_processor.iaudio_processor_iid);
    try printTuid(stdout, "IAudioPresentationLatency", audio_processor.iaudio_presentation_latency_iid);
    try printTuid(stdout, "IProcessContextRequirements", audio_processor.iprocess_context_requirements_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printOffset(writer: anytype, comptime type_name: []const u8, comptime field_label: []const u8, comptime Type: type, comptime field_name: []const u8) !void {
    try writer.print("{s}.{s} offset {}\n", .{ type_name, field_label, @offsetOf(Type, field_name) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}

const EventContext = struct {
    offset_sum: base.int32 = 0,
};

fn collectEvent(context: *EventContext, event: *const events.Event) void {
    context.offset_sum += event.sampleOffset;
}

const ParamContext = struct {
    param_id: vsttypes.ParamID = 0,
    offset_sum: base.int32 = 0,
    value_sum: vsttypes.ParamValue = 0,
};

fn collectParamPoint(context: *ParamContext, param_id: vsttypes.ParamID, sample_offset: base.int32, value: vsttypes.ParamValue) void {
    context.param_id = param_id;
    context.offset_sum += sample_offset;
    context.value_sum += value;
}

const LastParamContext = struct {
    offset: base.int32 = 0,
    value: vsttypes.ParamValue = 0,
};

fn collectLastParamPoint(context: *LastParamContext, _: vsttypes.ParamID, sample_offset: base.int32, value: vsttypes.ParamValue) void {
    context.offset = sample_offset;
    context.value = value;
}

const ChangesContext = struct {
    param_id_sum: vsttypes.ParamID = 0,
};

fn collectParamQueue(context: *ChangesContext, queue: *parameter_changes.IParamValueQueue) void {
    context.param_id_sum += queue.vtable.getParameterId(queue);
}

const MockEventList = struct {
    iface: events.IEventList,
    event_storage: [3]events.Event,

    fn init() MockEventList {
        return .{
            .iface = .{ .vtable = &event_list_vtable },
            .event_storage = .{
                .{ .sampleOffset = 10, .type = @intFromEnum(events.Event.EventTypes.kNoteOnEvent) },
                .{ .sampleOffset = 20, .type = @intFromEnum(events.Event.EventTypes.kNoteOffEvent) },
                .{ .sampleOffset = 30, .type = @intFromEnum(events.Event.EventTypes.kDataEvent) },
            },
        };
    }

    const event_list_vtable = events.IEventListVTable{
        .queryInterface = eventListQueryInterface,
        .addRef = eventListAddRef,
        .release = eventListRelease,
        .getEventCount = eventListGetEventCount,
        .getEvent = eventListGetEvent,
        .addEvent = eventListAddEvent,
    };
};

fn eventListSelf(self: *anyopaque) *MockEventList {
    return @ptrCast(@alignCast(self));
}

fn eventListQueryInterface(_: *anyopaque, _: *const @import("zig-vst3").tuid.TUID, obj: *?*anyopaque) callconv(.C) base.tresult {
    obj.* = null;
    return base.kNoInterface;
}

fn eventListAddRef(_: *anyopaque) callconv(.C) base.uint32 {
    return 1;
}

fn eventListRelease(_: *anyopaque) callconv(.C) base.uint32 {
    return 1;
}

fn eventListGetEventCount(_: *anyopaque) callconv(.C) base.int32 {
    return 3;
}

fn eventListGetEvent(self: *anyopaque, index: base.int32, event: *events.Event) callconv(.C) base.tresult {
    if (index == 1) return base.kResultFalse;
    event.* = eventListSelf(self).event_storage[@intCast(index)];
    return base.kResultOk;
}

fn eventListAddEvent(_: *anyopaque, _: *events.Event) callconv(.C) base.tresult {
    return base.kResultFalse;
}

const MockParamValueQueue = struct {
    iface: parameter_changes.IParamValueQueue,
    param_id: vsttypes.ParamID,
    sample_offsets: [3]base.int32 = .{ 4, 8, 12 },
    values: [3]vsttypes.ParamValue = .{ 0.25, 0.5, 0.75 },

    fn init(param_id: vsttypes.ParamID) MockParamValueQueue {
        return .{
            .iface = .{ .vtable = &param_queue_vtable },
            .param_id = param_id,
        };
    }

    const param_queue_vtable = parameter_changes.IParamValueQueueVTable{
        .queryInterface = paramQueueQueryInterface,
        .addRef = paramQueueAddRef,
        .release = paramQueueRelease,
        .getParameterId = paramQueueGetParameterId,
        .getPointCount = paramQueueGetPointCount,
        .getPoint = paramQueueGetPoint,
        .addPoint = paramQueueAddPoint,
    };
};

fn paramQueueSelf(self: *anyopaque) *MockParamValueQueue {
    return @ptrCast(@alignCast(self));
}

fn paramQueueQueryInterface(_: *anyopaque, _: *const @import("zig-vst3").tuid.TUID, obj: *?*anyopaque) callconv(.C) base.tresult {
    obj.* = null;
    return base.kNoInterface;
}

fn paramQueueAddRef(_: *anyopaque) callconv(.C) base.uint32 {
    return 1;
}

fn paramQueueRelease(_: *anyopaque) callconv(.C) base.uint32 {
    return 1;
}

fn paramQueueGetParameterId(self: *anyopaque) callconv(.C) vsttypes.ParamID {
    return paramQueueSelf(self).param_id;
}

fn paramQueueGetPointCount(_: *anyopaque) callconv(.C) base.int32 {
    return 3;
}

fn paramQueueGetPoint(self: *anyopaque, index: base.int32, sample_offset: *base.int32, value: *vsttypes.ParamValue) callconv(.C) base.tresult {
    if (index == 1) return base.kResultFalse;
    const queue = paramQueueSelf(self);
    sample_offset.* = queue.sample_offsets[@intCast(index)];
    value.* = queue.values[@intCast(index)];
    return base.kResultOk;
}

fn paramQueueAddPoint(_: *anyopaque, _: base.int32, _: vsttypes.ParamValue, _: *base.int32) callconv(.C) base.tresult {
    return base.kResultFalse;
}

const MockParameterChanges = struct {
    iface: parameter_changes.IParameterChanges,
    queues: [2]MockParamValueQueue,

    fn init() MockParameterChanges {
        return .{
            .iface = .{ .vtable = &parameter_changes_vtable },
            .queues = .{
                MockParamValueQueue.init(1234),
                MockParamValueQueue.init(5678),
            },
        };
    }

    const parameter_changes_vtable = parameter_changes.IParameterChangesVTable{
        .queryInterface = parameterChangesQueryInterface,
        .addRef = parameterChangesAddRef,
        .release = parameterChangesRelease,
        .getParameterCount = parameterChangesGetParameterCount,
        .getParameterData = parameterChangesGetParameterData,
        .addParameterData = parameterChangesAddParameterData,
    };
};

fn parameterChangesSelf(self: *anyopaque) *MockParameterChanges {
    return @ptrCast(@alignCast(self));
}

fn parameterChangesQueryInterface(_: *anyopaque, _: *const @import("zig-vst3").tuid.TUID, obj: *?*anyopaque) callconv(.C) base.tresult {
    obj.* = null;
    return base.kNoInterface;
}

fn parameterChangesAddRef(_: *anyopaque) callconv(.C) base.uint32 {
    return 1;
}

fn parameterChangesRelease(_: *anyopaque) callconv(.C) base.uint32 {
    return 1;
}

fn parameterChangesGetParameterCount(_: *anyopaque) callconv(.C) base.int32 {
    return 3;
}

fn parameterChangesGetParameterData(self: *anyopaque, index: base.int32) callconv(.C) ?*parameter_changes.IParamValueQueue {
    if (index == 1) return null;
    const changes = parameterChangesSelf(self);
    return &changes.queues[if (index == 0) 0 else 1].iface;
}

fn parameterChangesAddParameterData(_: *anyopaque, _: *const vsttypes.ParamID, _: *base.int32) callconv(.C) ?*parameter_changes.IParamValueQueue {
    return null;
}
